tape/src/main/java/com/squareup/tape/QueueFile.java

```diff
@@ -457,6 +467,68 @@ public synchronized int forEach(ElementVisitor reader) throws IOException {
     return elementCount;
   }
 
+  @Override public Iterator<byte[]> iterator() {
+    return new ElementIterator();
+  }
+
+  private final class ElementIterator implements Iterator<byte[]> {
+    /** Index of element to be returned by subsequent call to next. */
+    int nextElementIndex = 0;
+
+    /** Position of element to be returned by subsequent call to next. */
+    private int nextElementPosition = first.position;
+
+    /**
+     * The {@link #modCount} value that the iterator believes that the backing QueueFile should
+     * have. If this expectation is violated, the iterator has detected concurrent modification.
+     */
+    int expectedModCount = modCount;
+
+    @Override public boolean hasNext() {
+      return nextElementIndex != elementCount;
+    }
+
+    @Override public byte[] next() {
+      if (modCount != expectedModCount) throw new ConcurrentModificationException();
+      if (isEmpty()) throw new NoSuchElementException();
+
+      try {
+        // Read the current element.
+        Element current = readElement(nextElementPosition);
+        byte[] buffer = new byte[current.length];
+        nextElementPosition = wrapPosition(current.position + Element.HEADER_LENGTH);
+        ringRead(nextElementPosition, buffer, 0, current.length);
+
+        // Update the pointer to the next element.
+        nextElementPosition =
+            wrapPosition(current.position + Element.HEADER_LENGTH + current.length);
+        nextElementIndex++;
+
+        // Return the read element.
+        return buffer;
+      } catch (IOException e) {
+        throw new RuntimeException("todo: throw a proper error", e);
```

Comment:
Unsure what to throw here.

I think the options are:

1. throw FileException |  IOError | RuntimeException
2. Return an Object with an `InputStream` and length, instead of `byte[]` https://github.com/square/tape/pull/72#issuecomment-73346088.